Intial implementation of display for arguments to function call

sivaramaaa opened the pull request:
hey i have implemented an display for **arguments** to function call in visual mode , which imho could be useful instead of checking value in stack and register  every time  

#### To be able to view this : 
* enable this config var ``` e dbg.funcarg = true```
* And step into call instr in visual mode (works both while debugging and emulation)
#### Things to do for later : 
* Support display for arguments passed to function via syscall
* Improve function argument and type analysis 

![final](https://user-images.githubusercontent.com/18589720/36342746-e16d7722-1428-11e8-9407-7158509f368f.jpg)

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+	if (r_anal_type_func_exist (anal, func_name)) {
+		return strdup (func_name);
+	}
+	name = func_name;
```
define the var here, to LOC--

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+                break;
+	case 'd' : // integer
+	case 'x' :
+                r_cons_printf ("0x%08" PFMT64x, bval);
```
spaces here

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+		break;
+		}
+        }
+	free (buf);
```
oops more spaces vs tabs

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+	case 'p' : // pointer
+		{
+		// Try to deref the pointer once again
+		ut64 ptr_val = get_buf_val(buf, endian, width);
```
missing space

radare replied:
no need to hold this variable, can be passed directly to printf

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+/* This functon display list of arg with some colors */
+
+R_API void r_core_print_func_args(RCore *core) {
+	RAnalOp *op = {0};
```
thats actually NULL, not {0}, and can be defined in the line that is set

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+	if (src && !strcmp (src, "stack_rev")) {
+		for (i = nargs - 1; i >= 0; i--) {
+			RAnalFuncArg *arg = R_NEW0 (RAnalFuncArg);
+			set_fcn_args_info (arg, core->anal, key, cc, i);
```
possible null deref

sivaramaaa replied:
if u are taking about anal being null , i have added a check in set_fcn_args_info ()

radare replied:
im talking about arg being null, if R_NEW0 fails

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+        switch (opt) {
+        case 'z' : // Null terminated string
+                r_cons_print ("\"");
+		for (int i = 0; i < MAXSTRLEN; i++) {
```
spaces vs tabs

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+}
+static void print_format_values(RCore *core, const char *fmt, bool onstack, ut64 src) {
+        char opt;
+	ut64 bval = src;
```
spaces

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,255 @@
+	}
+	return r_anal_type_func_guess (anal, (char*)func_name);
+}
+static ut64 get_buf_val(ut8 *buf, int endian, int width) {
```
add empty line before function

radare commented on libr/core/carg.c:
```diff
@@ -0,0 +1,251 @@
+		}
+	} else {
+		for (i = 0; i < nargs; i++) {
+			RAnalFuncArg *arg = R_NEW0 (RAnalFuncArg);
```
chk for null in arg

radare commented:
lgtm

radare commented:
what's the difference between afal and afvd ?

radare commented:
i see, afvd works in debugger only and your command works in both. wonder if there are more differences in the implementations than this. i can merge and we can compare and discuss later which is the better command to do that (name and impl)

XVilka commented:
My impression about visual representations. I don't think those "-----" lines are really required. In my opinion those just takes 2 previous lines of your terminal. I would think about just differentiating them by color.

sivaramaaa commented:
Ok cool , i could work on that !!!

XVilka commented on libr/core/carg.c:
```diff
@@ -0,0 +1,248 @@
+	} else {
+		arg->c_type = arg->orig_c_type;
+	}
+	const char *query = sdb_fmt (-1, "type.%s", arg->c_type);
```
Keep in mind there are can be a complex types like structs and unions. Even if you do not display them - you need to handle those cases correctly still.

sivaramaaa replied:
Yup lot of work to do , also function argument analysis needs more improvement ,  i have a few in my mind , i could do that whenever time permits , that's why for now i have disabled output by default and u have to use the config var 'dgb.funcarg' to enable it

XVilka commented on libr/core/carg.c:
```diff
@@ -0,0 +1,248 @@
+	return (width == 8)? r_read_ble64 (buf, endian) : (ut64) r_read_ble32 (buf,endian);
+}
+
+static void print_format_values(RCore *core, const char *fmt, bool onstack, ut64 src) {
```
I think this function can be generalized and moved somewhere in libr/util or so

sivaramaaa replied:
There's already a function to print according to given format ('pf' command) , maybe we can generalize that , it's just that i didin't like the output of that , so implemented something minimalistic , so that i could tweak to my needs (like printing '-->' for pointers and strings)

sivaramaaa commented on libr/core/carg.c:
```diff
@@ -0,0 +1,248 @@
+			}
+			if (nargs > 0) {
+				for (i = 0; i < nargs; i++) {
+					ut64 v = r_debug_arg_get (core->dbg, R_ANAL_CC_TYPE_STDCALL, i);
```
No , this is done as worst-case scenario , see the func r_core_get_func_args() it fetches the arg depending on arch and function calling convention this is done only if that function fails  (also note these functions were taken out from disas.c and i felt that it's better to display something instead of void)

Maijin replied:
yep

radare commented:
Please resolve conflicts and finish the pr to merge it

sivaramaaa commented:
Just a sec i have one more commit to do , cleaning up the code

sivaramaaa commented:
I have removed the dotted lines and added a few color to differentiate as suggested by @XVilka 
In future i would make more improvements and try to remove the hard-coded color and support themes etc .. (But that's for another PR)
 
Updated Output : 

![new](https://user-images.githubusercontent.com/18589720/36465217-21fa9dd2-16fa-11e8-836a-5ccc6cd46297.png)

And yep this is it for this PR !!!

XVilka commented:
@sivaramaaa can you please rebase and squash?

sivaramaaa commented:
Yup done !!!

XVilka commented:
@sivaramaaa seems it breaks `pdc`:
```
[XX] db/cmd/cmd_pdc pdc @ main ioli0x0  35
```

sivaramaaa commented:
No , imo the current output is the correct one that's why i have updated the test and sent a PR in r2r , it's due change in stack pointer value , it was not updated proprly before !!!

XVilka commented:
@sivaramaaa can you please add a page about Visual Debugging mode in this chapter? https://radare.gitbooks.io/radare2book/content/visual_mode/intro.html
At least with your new feature

sivaramaaa commented:
Yeah sure , will do !!!
